contracts/stake-cw20-external-rewards/src/contract.rs

```diff
@@ -0,0 +1,267 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{to_binary, Addr, Binary, CosmosMsg, Deps, DepsMut, Env, MessageInfo, Response, StdError, StdResult, Storage, Uint128, WasmMsg, BankMsg, Coin, from_binary};
+use cw2::set_contract_version;
+use stake_cw20::hooks::StakeChangedHookMsg;
+use std::borrow::Borrow;
+use std::cmp::{max, min};
+use std::io::Empty;
+use cosmwasm_std::OverflowOperation::Add;
+use cw20::{Cw20ReceiveMsg, Denom};
+
+const CONTRACT_NAME: &str = "crates.io:stake_cw20";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response<Empty>, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let admin = match msg.admin {
+        Some(admin) => Some(deps.api.addr_validate(admin.as_str())?),
+        None => None,
+    };
+
+    let config = Config {
+        admin,
+        staking_contract: msg.staking_contract,
+        reward_token: msg.reward_token,
+    };
+    CONFIG.save(deps.storage, &config)?;
+    
+    let reward_config = RewardConfig{
+        periodFinish: 0,
+        rewardRate: Default::default(),
+        rewardDuration: 100000
+    };
+    REWARD_CONFIG.save(deps.storage, &reward_config);
+
+    Ok(Response::new())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response<Empty>, ContractError> {
+    match msg {
+        ExecuteMsg::StakeChangeHook(msg) => execute_stake_changed(deps, env, info, msg),
+        ExecuteMsg::Claim {} => execute_claim(deps, env, info),
+        ExecuteMsg::Fund {} => execute_fund_native(deps, env, info),
+        ExecuteMsg::Receive(msg) => execute_receive(deps, env, info, msg)
+    }
+}
+
+pub fn execute_receive(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    wrapper: Cw20ReceiveMsg
+) -> Result<Response<Empty>, ContractError> {
+    let msg: ReceiveMsg = from_binary(&wrapper.msg)?;
+    let config = CONFIG.load(deps.storage)?;
+    let sender = deps.api.addr_validate(&*wrapper.sender)?;
+    if config.reward_token != Denom::Cw20(info.sender) {
+        return Err(Unauthorized {})
+    };
+    if config.admin != Some(sender.clone()){
+        return Err(Unauthorized {})
+    };
+    match msg {
+        ReceiveMsg::Fund { .. } => {execute_fund(deps, env, sender, wrapper.amount)}
+    }
+}
+
+pub fn execute_fund_native (
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo
+) -> Result<Response<Empty>, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if config.admin != Some(info.sender.clone()){
+        return Err(Unauthorized {})
+    };
+    let coin = info.funds.first()?;
+    let amount = coin.clone().amount;
+    let denom = coin.clone().denom;
```

Comment:
```suggestion
    let { amount, denom } = coin;
```

Reply:
hmm, i think this suggestion is invalid. not sure proper syntax here to do the destructuring